auth/common.cc

```diff
@@ -113,4 +118,118 @@ ::service::query_state& internal_distributed_query_state() noexcept {
     return qs;
 }
 
+future<::service::group0_guard> start_group0_operation(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        seastar::abort_source* as) {
+    co_return co_await qp.container().invoke_on(0,
```

Comment:
I am also not sure that it is safe to use `abort_source` across shards.

Reply:
> Will also prevent cross cpu freeing of the internal grou0_pointer.

There is no cross shard free the guard is moved back to shard0 and freed there, but the jumping between shards is redundant. Everything, including the data, is in the shard0.

Reply:
May be we can bounce entire query to shard0. See `result_message::bounce_to_shard`.

Reply:
for jumping to shard 0 earlier it's doable but it'd mean that almost the whole code (selects, generating mutations, etc) is executed on shard 0, I am not sure do we want that?

as for abort_source, looks like indeed it the one from apporpiate shard should be used.

Reply:
> for jumping to shard 0 earlier it's doable but it'd mean that almost the whole code (selects, generating mutations, etc) is executed on shard 0, I am not sure do we want that?
> 

What is the problem with it? And select will have to jump to shard0 anyway since all the data is there.

Reply:
btw if we decide to jump to shard0 early in query_processor (on execution of AUTH DML  not internals ones) that would solve both, with mentioned shard0 congestion caveat

Reply:
Yes.

Reply:
> What is the problem with it? And select will have to jump to shard0 anyway since all the data is there.

Are you sure? I think group0 underneath calls `static future<> write_mutations_to_database` which then does `mutate_locally` and it applies in parallel on the owning shards.

Reply:
> > What is the problem with it? And select will have to jump to shard0 anyway since all the data is there.
> 
> Are you sure? I think group0 underneath calls `static future<> write_mutations_to_database` which then does `mutate_locally` and it applies in parallel on the owning shards.

All tables managed by group0 are usually on shard0. Do they have to be? May be not, but then you need to be careful to transfer mutations from all shards in your snapshot transfer code (I did not see it yet). Also if you want serialized reads you need to run raft barrier  which can be done only from shard0. But without serialized reads taking care of correct snapshot transfer should probably work.

But again all such tables are on shard0 right now so who knows what will break otherwise. For instance we have https://github.com/scylladb/scylladb/issues/15596 where we want to move all tables managed by raft to be in schema commitlog and this one exists only on shard0.

Reply:
Removed those invoke_on by bouncing to shard0 with needs_guard=true, abort_source problem fixed itself as a result.

One caveat is that I am not fully using this guard from query_processor, the reason is auth code like this:
```c++
future<> create_role(
        const service& ser,
        std::string_view name,
        const role_config& config,
        const authentication_options& options) {
    return ser.underlying_role_manager().create(name, config).then([&ser, name, &options] {
        if (!auth::any_authentication_options(options)) {
            return make_ready_future<>();
        }

        return futurize_invoke(
                &validate_authentication_options_are_supported,
                options,
                ser.underlying_authenticator().supported_options()).then([&ser, name, &options] {
            return ser.underlying_authenticator().create(name, options);
        }).handle_exception([&ser, name](std::exception_ptr ep) {
            // Roll-back.
            return ser.underlying_role_manager().drop(name).then([ep = std::move(ep)] {
                std::rethrow_exception(ep);
            });
        });
    });
}
```

One call is composed out of multiple lower level calls each doing group0 ops and once I move guard to one of them I no loner have it for other calls. 

One way to address this is to rewrite the whole code to generate mutations and announce them at the end but as you can see there is also "roll-back" for reason I don't understand - probably not needed with auth-v2. Such rewrite in my view should be out of scope for now because I don't think landing this change combined with rewrite of half of auth code will be a great success :)

Reply:
The rewrite you mentioned is exactly the kind of rewrite that was done to move schema statements to raft. It will definitely complicate the patch series, but without it do we really have "linearizable write" guaranty that we what from the move to raft?

Reply:
We don't have such guaranty but we also don't need it to say that new implementation is as good as old (eventually consistent) one. And in fact it's much better because the main reason to work on this is not even consistency but data replication - so that data is replicated everywhere and each read is local.

Reply:
It is very troubling if we start using raft as "replicate everywhere" strategy. And I think we do want to linearize auth with DDL statements which is the main point of the move.

Reply:
> auth with DDL statements which is the main point of the move

how do you deduce that this is the main point?

Reply:
I believe the comment is already addressed, are you sure you want to continue the argument?

Reply:
> I believe the comment is already addressed, are you sure you want to continue the argument?

Which comment is addressed?